crates/matrix-sdk-base/src/read_receipts.rs

```diff
@@ -0,0 +1,395 @@
+use matrix_sdk_common::deserialized_responses::SyncTimelineEvent;
+use ruma::{
+    events::{
+        poll::{start::PollStartEventContent, unstable_start::UnstablePollStartEventContent},
+        receipt::{ReceiptThread, ReceiptType},
+        room::message::Relation,
+        AnySyncMessageLikeEvent, AnySyncTimelineEvent, OriginalSyncMessageLikeEvent,
+        SyncMessageLikeEvent,
+    },
+    serde::Raw,
+    EventId, OwnedEventId, RoomId, UserId,
+};
+use tracing::{field::display, instrument, trace};
+
+use super::BaseClient;
+use crate::{error::Result, store::StateChanges, RoomInfo};
+
+/// Provider for timeline events prior to the current sync.
+pub trait PreviousEventsProvider: Send + Sync {
+    /// Returns the list of known timeline events, in sync order, for the given
+    /// room.
+    // TODO: return a reference or some kind of iterator
+    fn for_room(&self, room_id: &RoomId) -> Vec<SyncTimelineEvent>;
+}
+
+impl PreviousEventsProvider for () {
+    fn for_room(&self, _: &RoomId) -> Vec<SyncTimelineEvent> {
+        Vec::new()
+    }
+}
+
+#[instrument(skip_all, fields(room_id))]
+pub(crate) async fn compute_notifications(
+    client: &BaseClient,
+    changes: &StateChanges,
+    previous_events_provider: &dyn PreviousEventsProvider,
```

Comment:
Maybe add `PreviousEventsProvider` as a generic parameter, so that we get static dispatch instead of a dynamic dispatch here? I don't see which pros dynamic dispatch brings here:

```rust
pub(crate) async fn compute_notifications<P = ()>(
    clients: &BaseClient,
    changes: &StateChanges,
    previous_events_provider: &P,
    new_events: &[SyncTimelineEvent],
    room_info: &mut RoomInfo,
) -> Result<()>
where
    P: PreviousEventsProvider
{
```

It also has the advantage of using a default value, which simplifies calls, particularly in the tests.

Reply:
Ah yeah, old habit of trying to avoid generics, but you're right it should work nicely here.

> It also has the advantage of using a default value, which simplifies calls, particularly in the tests.

We still need to specify a parameter, so it doesn't change anything for the tests.